src/Datadog.Trace.ClrProfiler.Managed/CallTarget/Handlers/IntegrationMapper.cs

```diff
@@ -0,0 +1,714 @@
+                {
+                    throw new InvalidCastException($"The target parameter {targetParameterType} can't be assigned from {sourceParameterType}");
+                }
+
+                ILHelpersExtensions.WriteLoadArgument(ilWriter, i, mustLoadInstance);
+                if (parameterProxyType != null)
+                {
+                    ConstructorInfo parameterProxyTypeCtor = parameterProxyType.GetConstructors()[0];
+
+                    if (sourceParameterType.IsValueType && !parameterProxyTypeCtor.GetParameters()[0].ParameterType.IsValueType)
+                    {
+                        ilWriter.Emit(OpCodes.Box, sourceParameterType);
+                    }
+
+                    ilWriter.Emit(OpCodes.Newobj, parameterProxyTypeCtor);
+                }
+            }
+
+            // Call method
+            // Log.Information("Generic Types: " + string.Join(", ", callGenericTypes.Select(t => t.FullName)));
+            // Log.Information("Method: " + onMethodBeginMethodInfo);
+            onMethodBeginMethodInfo = onMethodBeginMethodInfo.MakeGenericMethod(callGenericTypes.ToArray());
+            // Log.Information("Method: " + onMethodBeginMethodInfo);
+            ilWriter.EmitCall(OpCodes.Call, onMethodBeginMethodInfo, null);
+            ilWriter.Emit(OpCodes.Ret);
+
+            Log.Debug($"Created BeginMethod Dynamic Method for '{integrationType.FullName}' integration. [Target={targetType.FullName}]");
+            return callMethod;
+        }
+
+        internal static DynamicMethod CreateSlowBeginMethodDelegate(Type integrationType, Type targetType)
+        {
+            Log.Debug($"Creating SlowBeginMethod Dynamic Method for '{integrationType.FullName}' integration. [Target={targetType.FullName}]");
+            MethodInfo onMethodBeginMethodInfo = integrationType.GetMethod(BeginMethodName, BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Static);
+            if (onMethodBeginMethodInfo is null)
+            {
+                throw new NullReferenceException($"Couldn't find the method: {BeginMethodName} in type: {integrationType.FullName}");
+            }
+
+            if (onMethodBeginMethodInfo.ReturnType != typeof(CallTargetState))
+            {
+                throw new ArgumentException($"The return type of the method: {BeginMethodName} in type: {integrationType.FullName} is not {nameof(CallTargetState)}");
+            }
+
+            Type[] genericArgumentsTypes = onMethodBeginMethodInfo.GetGenericArguments();
+            if (genericArgumentsTypes.Length < 1)
+            {
+                throw new ArgumentException($"The method: {BeginMethodName} in type: {integrationType.FullName} doesn't have the generic type for the instance type.");
+            }
+
+            ParameterInfo[] onMethodBeginParameters = onMethodBeginMethodInfo.GetParameters();
```

Comment:
Similar to the `CreateBeginMethodDelegate` method body, should we add a hard check to make sure that if we define an instance method (first argument with type `TTarget`, second argument with type `object[]`) the first method parameter has the same generic type as the generic method?

```suggestion
            ParameterInfo[] onMethodBeginParameters = onMethodBeginMethodInfo.GetParameters();
            if (onMethodBeginParameters.Length > 1 && onMethodBeginParameters[0].ParameterType != genericArgumentsTypes[0])
            {
                throw new ArgumentException($"The first generic argument for method: {BeginMethodName} in type: {integrationType.FullName} must be the same as the first parameter for the instance value.");
            }
```

Reply:
I'll resolve this note so as not to lose history, but I think I was confused. `onMethodBeginMethodInfo` is our integration method which **will not** have just two parameters with parameter0=`TTarget` and parameter1=`object[]`. The integration method will have a bunch of strongly-typed or proxy-typed parameters and this dynamic method unpacks the `object[]` into the long list of arguments needed by the integration method. As a result, this check I was suggesting is unnecessary because this will misbehave if the integration method is static.